mountpoint-s3-client/src/s3_crt_client/put_object.rs

```diff
@@ -45,19 +47,53 @@ impl S3CrtClient {
                 .set_header(&Header::new("x-amz-storage-class", storage_class))
                 .map_err(S3RequestError::construction_failure)?;
         }
-
+        let mut maybe_sse = None;
+        let mut maybe_key_id = None;
+        match params.server_side_encryption.to_owned() {
+            ServerSideEncryption::Default => (),
+            ServerSideEncryption::Kms { key_id } => {
+                maybe_sse = Some("aws:kms");
+                maybe_key_id = key_id;
+            }
+            ServerSideEncryption::DualLayerKms { key_id } => {
+                maybe_sse = Some("aws:kms:dsse");
+                maybe_key_id = key_id;
+            }
+        }
+        let mut expected_headers = Vec::new();
+        if let Some(sse) = maybe_sse {
+            let header_name = "x-amz-server-side-encryption";
+            expected_headers.push((header_name.to_owned(), sse.to_owned()));
+            message
+                .set_header(&Header::new(header_name, sse))
+                .map_err(S3RequestError::construction_failure)?;
+        }
+        if let Some(key_id) = maybe_key_id {
+            let header_name = "x-amz-server-side-encryption-aws-kms-key-id";
+            expected_headers.push((header_name.to_owned(), key_id.clone()));
+            message
+                .set_header(&Header::new(header_name, key_id))
+                .map_err(S3RequestError::construction_failure)?;
+        }
+        let response_headers: Arc<Mutex<Option<Headers>>> = Default::default();
+        let response_headers_writer = response_headers.clone();
```

Comment:
that was helpful: https://huonw.github.io/blog/2015/05/finding-closure-in-rust/ (in particular thinking of closures as structs containing the captured environment)

from what I understand:
- `response_headers_writer` is owned by the closure and will get dropped, when the closure is dropped **(not necessarily after first invocation of a closure**, since it is `FnMut`)
- closure will eventually get owned by `MetaRequestOptions` and, thus, closure is dropped at the same time as `MetaRequestOptions` which **will** happen at some point

resolving